scripts/bench_summary.lean

```diff
@@ -0,0 +1,184 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean.Elab.Command
+
+/-!
+# Summary of `!bench` results
+
+This file contains a script that converts data retrieved from the speed-center into a
+shorter, more accessible format.
+-/
+
+namespace BenchAction
+
+open Lean
+
+/-- `Bench` is a structure with the data used to compute the `!bench` summary.
+It contains
+* a string `file` (that could be `build`, `lint` or `~Mathlib.Path.To.File`);
+* an integer `diff` representing the change in number of instructions for `file`;
+* a float `reldiff` representing the percentage change in number of instructions for `file`.
+-/
+structure Bench :=
+  file    : String
+  diff    : Int
+  reldiff : Float
+  deriving FromJson, ToJson, Inhabited
+
+/-- `intDecs z exp prec` is a "generic" formatting of an integer `z`.
+It writes the number as `x.y * 10 ^ expr`, where `y` has `prec` digits and returns
+* the sign of `z` as a string (in fact, just either `+` or `-`);
+* the integer `x`;
+* the natural number `y` (that has `prec` digits).
+-/
+def intDecs (z : Int) (exp : Nat := 9) (prec : Nat := 3) : String × Int × Nat :=
+  let sgn := z.sign
+  let z := sgn * z
+  let p10 : Int := 10 ^ (exp - prec)
+  let idec := z / p10
+  (if sgn < 0 then "-" else "+", idec / (10 ^ prec), (idec % 10 ^ prec).toNat)
+
+/-- `formatDiff z` uses `intDecs` to format the integer `z` as `±x.y⬝10⁹`. -/
+def formatDiff (z : Int) : String :=
+  let (sgn, intDigs, decDigs) := intDecs z
+  s!"{sgn}{intDigs}.{decDigs}⬝10⁹"
+
+/-- converts a `Float` into a formatted string of the form `±z.w%`. -/
+def formatPercent (reldiff : Float) : String :=
+  -- shift by `2` twice: the first one, to get a `%`; the second, for 2 decimal digits of precision
+  let reldiff := reldiff * 10 ^ 4
+  let (sgn : Int) := if reldiff < 0 then -1 else 1
+  let (sgnf : Float) := .ofInt sgn
+  let reldiff := sgnf * reldiff
+  let (sgn, intDigs, decDigs) := intDecs (sgn * reldiff.toUInt32.val) 0 2
+  s!"({sgn}{intDigs}.{decDigs}%)"
+
+/-- info: #["(+0.0%)", "(+14.28%)", "(+0.20%)", "(-0.60%)"] -/
```

Comment:
Great idea to add a test!

Reply:
Would you like to also test `formatDiff` (can be with the same data)?

Reply:
Added a test of `formatDiff` as well.